### ISSUE TITLE
Provide a standard directory for forc build artifacts / compiler output (like Rust's `target/` directory)

### ISSUE THREAD

**Issue by mitchmindtree** [FuelLabs/sway#765]
Currently, it doesn't seem like we have a standard directory for build artifacts produced by `forc` - please correct me if I'm wrong!

## Motivation

**`script` and `predicate` binaries**

It would be nice if we had a standard location to point users to when generating output from the compiler. Currently we can manually create a binary with `-o` and decide where to put it ourselves, but it would be nice if there was a default location where `forc build` could output this for `script`s and `predicate`s.

**`contract` ABI JSON**

E.g. the sway book [currently recommends](https://fuellabs.github.io/sway/latest/getting-started/forc_project.html) manually piping contract ABI JSON files to a project's root directory, but mentions we're yet to decide on a standard location for these.

This is also a bit of a problem for the test harnesses right now (see #764). it seems we don't yet have a standard location to point to when generating Rust bindings for contract ABIs, e.g.
```rust
abigen!(MyContract, "./my-contract-abi.json"); // Currently assumes project root.
```

## Prior Work

**Rust** uses a `<project_root>/target` directory. Calls to `cargo build` or `cargo build --release` in the project root will generate the `target` directory if it doesn't yet exist. A subdirectory is added for the build profile used (i.e. `target/debug` for `cargo build`, `target/release` for `cargo build --release`). The build artifacts (libraries, incremental compilation artifacts, binaries) are placed within the `target/<build-profile>/` directory.

**Foundry** seems to use a `<project_root>/out` directory. A call to `forge build` will generate the `out` directory if it doesn't yet exist. Here it places contract ABI output, but I haven't dove deep enough to know if it also places EVM bytecode here.

## Potential Solution

Like Rust and Foundry, it could be a good idea for `forc build` (and `forc test`) to output all build artifacts (contract ABIs, script/predicate binaries, any future caching artifacts) for the project into a standard output directory.

Seeing as we currently lean into Rust for test harnesses, we may want to avoid re-using the `target` directory to avoid any possibility of conflict between artifacts created by `forc` and `cargo`.

I think `out` (named after compiler **out**put) is a nice concise alternative that shouldn't conflict with `cargo` at all. That said, if we're to add this feature, the output directory should be configurable (e.g. `--output-directory`) to remain friendly/interoperable with other build tools like nix.

Seeing as a forc project can be only one of either a contract, predicate, script or library, and we don't currently distinguish between build profiles (e.g. debug, release, etc), it may be safe for us to just place artifacts directly at the top level of `out`.

## Implementation

Might look something like this:

1. Add a `--output-directory` option to `forc build` that takes a path to a directory. If unspecified, defaults to `<project-root>/out`.
2. On `forc build` create the output directory if it doesn't yet exist.
3. On `forc build`, depending on the project type, output the relevant build artifact to the output directory. E.g.
   - **contract**s
     ```
     out/<project-name>-abi.json
     ```
   - **script**s and **predicate**s
     ```
     out/<project-name>.bin
     ```
   - **library**s could potentially omit the `out` directory altogether? Unless there are some debug / intermediate artifacts we may want to generate for them?
4. Update `forc test` to first invoke `forc build` to ensure any necessary contract ABIs exist.
5. Update example contract project test harnesses to generate Rust bindings for contract ABIs to the new default location:
   ```rust
   abigen!(MyContract, "./out/<project-name>-abi.json");
   ```
6. Update `forc init` to create a `.gitignore` containing `out`.
7. Add a `forc clean` command that just deletes the `out` directory.

## Alternatives

In retrospect, seeing as we would only be generating a single output file per project (as each project can only be one of contract, script, predicate or library), and we don't yet do any caching or have any incremental compilation artifacts, this might be overkill. There is arguably some benefit to having `forc` remain more stateless as its easier to maintain reproducibility.

The current issue with test harnesses having predictable access to build artifacts could alternatively be solved by:

1. Creating a temporary directory relative to the project root that is guaranteed only to live as long as `forc test` runs. 
2. At the beginning of `forc test`, call `forc build` and output the build artifact (binary or contract ABI) to this temporary directory.
3. Run tests as usual where `abigen!` retrieves contract ABI via something like:
   ```rust
   abigen!(MyContract, "./tmp/<project-name>-abi.json");
   ```
4. Deleting temporary directory before exiting.

---

I'm a little torn on what the right approach is here, any thoughts appreciated!

**Comment by adlerjohn** [comment]
Thoughts dump:

1. `out` seems reasonable.
2. Contracts should also output a bytecode binary file, just like scripts and predicates. So that means we'll have a case already where more than one output, and so an immediate need for a directory.
3. The `.gitignore` should probably add `target` as well? And likewise `forc clean` should delete `target` (maybe behind an `--all` or something).
4. The SDK macro should be split into `abigen!` and `abigen_from_file!`, with the first simply calling the latter with the default filename for the appropriate profile (may need some logic to get the profile automatically??? is this even possible?). Can be left to @digorithm
5. IMO we should start with `out/debug/` rather than `out/` just to avoid having to break paths later. The existence of a second profile can be deferred for later.

**Comment by mitchmindtree** [comment]
> The SDK macro should be split into abigen! and abigen_from_file!,

I opened this just recently on a related note https://github.com/FuelLabs/fuels-rs/issues/94

**Comment by mitchmindtree** [comment]
Closed via #796
